### PR TITLE
DeltaCAT Native Storage Improvements - Phase 1

### PR DESCRIPTION
**Pull request by pdames**
## Summary

First phase of refactoring, bug fixes, additional tests, and general cleanup for the DeltaCAT native storage implementation. This PR still does not reflect a finished DeltaCAT native storage implementation. Its primary purpose is to simply checkpoint a more stable foundation for future PRs to continue building off of.

One notable change has been made to table version identifiers to enforce a stricter naming convention of the form `"TableVersionName.1"` or simply `"1"`. This enforced inclusion of the table version's ordinal position as part of the table version identifier simplifies both optional CAS-based conflict detection for users committing a new table version (i.e., I only want to commit table version 2 if the last table version was 1) and resolution of the latest active table version on table lifecycle state update.

Another notable change has been made to replace `SchemaMap` with named subschemas in the existing `Schema` class, where each named subschema forms a logical grouping of arrow schema fields, and persists subschema name to field ID mappings in arrow schema metadata. This refactoring lets the existing metadata storage model capture both schema evolution events involving either inserts/updates/deletes of existing subschemas and adding/removing/updating fields within any individual subschema with relatively minimal code changes.

Also updates Rivulet print statements to instead use the DeltaCAT system logger with more appropriate verbosity levels.

## Checklist

- [X] Unit tests covering the changes have been added
  - [ ] If this is a bugfix, regression tests have been added

- [ ] E2E testing has been performed

### CODE REVIEW

File: deltacat/tests/storage/main/test_main_storage.py [hunk 1]
```diff
@@ -100,157 +84,318 @@ def test_namespace_exists_nonexisting(self):
...
+            namespace=cls.namespace.namespace,
             catalog=cls.catalog,
         )
+
         # Create a "base" table to attach versions to
-        metastore.create_table_version(
-            namespace="test_tv_ns",
-            table_name="mytable",
-            table_version="v1",
+        cls.stream1 = metastore.create_table_version(
```
Comment (mcember) [inline deltacat/tests/storage/main/test_main_storage.py:184]: should be a table version and not a stream

File: deltacat/tests/storage/model/test_table_version.py [hunk 2]
```diff
@@ -1,26 +1,98 @@
...
+)
+def test_parse_version(table_version, expected_parsed_version):
+    prefix, version_number = TableVersion.parse_table_version(table_version)
+    if prefix is not None:
+        assert isinstance(prefix, str)
+    assert isinstance(version_number, int)
+    assert (prefix, version_number) == expected_parsed_version
+
+
+def test_version_validation():
```
Comment (mcember) [inline deltacat/tests/storage/model/test_table_version.py:40]: nitpick: there are multiple cases here that could be separate functions (e.g. testing leading 0 truncation)

**Review by mcember — APPROVED** [review]
Left a few minor comments. Approved pending you look at comments and address what you think is necessary

File: deltacat/storage/model/table_version.py [hunk 3]
```diff
@@ -3,16 +3,15 @@
...
+from typing import Any, Dict, List, Optional, Union, Tuple
 
 import pyarrow
 import pyarrow as pa
 
 import deltacat.storage.model.partition as partition
 
 from deltacat.storage.model.metafile import Metafile, MetafileRevisionInfo
-from deltacat.constants import TXN_DIR_NAME
+from deltacat.constants import TXN_DIR_NAME, BYTES_PER_KIBIBYTE
```
[the remark this answers is not in the document]
Comment (pdames) [reply deltacat/storage/model/table_version.py:14]: Kibibyte (KiB) or something else? If Kibibyte, we're using it intentionally here: https://dictionary.cambridge.org/us/dictionary/english/kibibyte

File: deltacat/storage/interface.py [hunk 4]
```diff
@@ -384,23 +398,23 @@ def delete_stream(
...
-    Raises an error if the table version does not exist.
+    Raises an error if the stream does not exist.
     """
     raise NotImplementedError("delete_stream not implemented")
 
 
-def get_staged_stream(
+def get_stream_by_id(
     table_version_locator: TableVersionLocator,
     stream_id: str,
```
[the remark this answers is not in the document]
Comment (pdames) [reply deltacat/storage/interface.py:408]: I had considered this, but I think the behavior of that method becomes less clear when adding `stream_id` to `get_stream` because `table_version` now becomes required instead of `Optional` whenever `stream_id` is also specified, which would require a careful reading of the function documentation (and more careful ongoing maintenance of this documentation) about what a valid set of input parameters constitutes. As far as asking for a specific stream ID attached to an unknown latest active table version, this also didn't seem practical to me since resolving a stream's unique ID first requires knowing exactly which table version ID that stream is attached to, and asking for a hopeful match of a stream ID against an unknown table version seems to be asking for trouble.

File: deltacat/storage/model/table_version.py [hunk 5]
```diff
@@ -323,22 +322,67 @@ def from_serializable(
...
+
+    @staticmethod
+    def parse_table_version(table_version: str) -> Tuple[Optional[str], int]:
+        """
+        Parses a table version string into its prefix and version number.
+        Returns a tuple of the prefix and version number.
+        """
+        if not table_version:
+            raise ValueError(f"Table version to parse is undefined.")
+        if len(table_version) > BYTES_PER_KIBIBYTE:
```
[the remark this answers is not in the document]
Comment (pdames) [reply deltacat/storage/model/table_version.py:369]: This static method is currently invoked in the constructor for table version since it is also used by the setter for table version, which in turn is called by the constructor. I've left it here as a separate static method since we also use it in several locations (e.g., in update_table_version) to validate table version identifiers that don't have a parent table version object.

File: deltacat/storage/interface.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (mcember) [reply deltacat/storage/interface.py:408]: makes sense

File: deltacat/storage/model/table_version.py [hunk 6]
```diff
@@ -323,22 +322,67 @@ def from_serializable(
             self.locator.table_locator = table.locator
         return self
 
+    def current_version_number(self) -> Optional[int]:
+        """
+        Returns the current table version number as an integer, or None if
+        a table version has not yet been assigned.
```
[the remark this answers is not in the document]
Comment (mcember) [reply deltacat/storage/model/table_version.py:328]: ok. Gradual construction makes sense. At least changing the placeholder pattern would be a relatively minor refactor and eliminate some edge cases

File: deltacat/tests/storage/main/test_main_storage.py (hunk 1 above)
Comment (mcember) [reply deltacat/tests/storage/main/test_main_storage.py:184]: Yea, I like returning all of the objects which are created/fetched during creation as a convenience
